domain/application/state/storage.go

```diff
@@ -118,6 +135,226 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
 	return nil
 }
 
+type storageTemplate struct {
+	meta   charmStorage
+	params application.ApplicationStorageArg
+}
+
+// insertUnitStorage inserts the storage records need to record the intent
+// for the specified new unit and storage args. Records include:
+// - storage instance
+// - filesystem
+// - volume
+// - related attachment records
+// TODO(storage) - support attaching existing storage when adding a unit
+func (st *State) insertUnitStorage(
+	ctx context.Context, tx *sqlair.TX, modelType model.ModelType, storageParentDir string,
+	appUUID coreapplication.ID, unitUUID coreunit.UUID, netNodeUUID string,
+	args []application.ApplicationStorageArg, poolKinds map[string]storage.StorageKind,
+) error {
+
+	// Reduce the count of new storage created for each existing storage
+	// being attached.
+	// TODO(storage) - implement this when unit machine storage can be supported
+	// (includes ensureCharmStorageCountChange below)
+
+	templates := make([]storageTemplate, 0, len(args))
+	for _, arg := range args {
+		storageMeta, err := st.getApplicationCharmStorageByName(ctx, tx, appUUID, arg.Name)
+		if errors.Is(err, sqlair.ErrNoRows) {
```

Comment:
This smells. The fact that you are checking for this error type suggest you are programming based on implementation and not contracts. What happens when the implementation changes? Can we use an internal error type to represent this and the ambiguity goes away.

Reply:
It's an internal method. It's fine to reply on some knowledge of implementation in such cases. It was introduced purely to avoid having an even large chunk f logic in the calling function. I can add an error though.